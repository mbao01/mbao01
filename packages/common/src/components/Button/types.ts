import { type ButtonHTMLAttributes } from "react";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "./constants";

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof getButtonClasses> & {
    asChild?: boolean;
  };
