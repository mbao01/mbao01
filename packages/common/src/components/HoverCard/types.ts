import * as HoverCardPrimitive from "@radix-ui/react-hover-card";

export type HoverCardProps = React.ComponentPropsWithoutRef<
  typeof HoverCardPrimitive.Root
>;
