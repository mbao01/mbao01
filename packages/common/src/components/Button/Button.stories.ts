import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    variant: 'default',
    label: 'Button',
  },
};

export const PrimaryButton: Story = {
  args: {
    variant: 'primary',
    label: 'Button',
  },
};

export const WideButton: Story = {
  args: {
    wide: true,
    label: 'Wide Button',
  },
};

export const TinyButton: Story = {
  args: {
    size: 'xs',
    label: 'Tiny',
  },
};

export const OutlineButton: Story = {
  args: {
    outline: true,
    label: 'Click me!',
  },
};

export const DisabledButton: Story = {
  args: {
    disabled: true,
    label: 'Click me?',
  },
};

export const LoadingButton: Story = {
  args: {
    loading: true,
    label: 'Loading',
  },
};

export const BlockButton: Story = {
  args: {
    disabled: true,
    className: 'btn-block',
    label: 'Block Button',
  },
};
