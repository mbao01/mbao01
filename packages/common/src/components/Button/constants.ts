import { cva } from "../../libs";

export const getButtonClasses = cva("btn disabled:text-opacity-50", {
  variants: {
    variant: {
      accent: "btn-accent",
      default: "btn-default",
      error: "btn-error",
      ghost: "btn-ghost",
      info: "btn-info",
      link: "btn-link",
      neutral: "btn-neutral",
      primary: "btn-primary",
      secondary: "btn-secondary",
      success: "btn-success",
      warning: "btn-warning",
    },
    outline: {
      true: "btn-outline",
    },
    rounded: {
      xs: "rounded-xs",
      sm: "rounded-sm",
      md: "rounded-md",
      lg: "rounded-lg",
    },
    wide: {
      true: "btn-wide",
    },
    size: {
      xs: "btn-xs",
      sm: "btn-sm",
      md: "btn-md",
      lg: "btn-lg",
    },
    isLoading: {
      true: "relative overflow-hidden",
    },
  },
  compoundVariants: [
    {
      size: undefined,
      className: "min-h-fit h-10",
    },
    {
      rounded: undefined,
      className: "rounded-md",
    },
    {
      variant: undefined,
      outline: true,
      className: "border-neutral-content",
    },
  ],
});
