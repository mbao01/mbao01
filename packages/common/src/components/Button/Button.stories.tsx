import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "./Button";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    children: "Button",
  },
};

export const PrimaryButton: Story = {
  args: {
    variant: "primary",
    children: "Button",
  },
};

export const WideButton: Story = {
  args: {
    wide: true,
    children: "Wide Button",
  },
};

export const TinyButton: Story = {
  args: {
    size: "xs",
    children: "Tiny",
  },
};

export const OutlineButton: Story = {
  args: {
    outline: true,
    children: "Click me!",
  },
};

export const DisabledButton: Story = {
  args: {
    disabled: true,
    children: "Click me?",
  },
};

export const LoadingButton: Story = {
  args: {
    isLoading: true,
    children: "Loading",
  },
};

export const BlockButton: Story = {
  args: {
    disabled: true,
    className: "btn-block",
    children: "Block Button",
  },
};

export const SlotButton: Story = {
  args: {
    asChild: true,
    className: "btn-block",
    children: (
      <a href="#" rel="noopener">
        Link
      </a>
    ),
  },
};
