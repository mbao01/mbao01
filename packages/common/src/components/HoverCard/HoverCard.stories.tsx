import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react-vite";
import { Button } from "../Button";
import { HoverCard } from "./HoverCard";
import { type HoverCardProps } from "./types";

const withHoverCard = (_: StoryFn, context: StoryContext<HoverCardProps>) => {
  return (
    <HoverCard {...context.args}>
      <HoverCard.Trigger asChild>
        <Button variant="link">@nextjs</Button>
      </HoverCard.Trigger>
      <HoverCard.Content className="w-80">
        <div className="space-y-1">
          <h4 className="text-sm font-semibold">@nextjs</h4>
          <p className="text-sm">The React Framework - created and maintained by @vercel.</p>
          <div className="flex items-center pt-2">
            <span className="text-xs text-muted-foreground">Joined December 2021</span>
          </div>
        </div>
      </HoverCard.Content>
    </HoverCard>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: HoverCard,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withHoverCard],
} satisfies Meta<typeof HoverCard>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};

export const OpenedByDefault: Story = {
  args: {
    defaultOpen: true,
  },
};

export const WithOpenDelay: Story = {
  args: {
    openDelay: 3000,
  },
};

export const WithCloseDelay: Story = {
  args: {
    closeDelay: 3000,
  },
};
