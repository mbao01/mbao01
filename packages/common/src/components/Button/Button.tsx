import React from "react";
import { Slot, Slottable } from "@radix-ui/react-slot";
import { cn } from "../../utilities";
import { Loading } from "../Loading";
import { getButtonClasses } from "./constants";
import { type ButtonProps } from "./types";

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      asChild,
      className,
      outline,
      children,
      isLoading,
      variant,
      size,
      wide,
      rounded,
      disabled,
      ...props
    },
    ref
  ) => {
    const Comp = asChild ? Slot : "button";

    return (
      <Comp
        ref={ref}
        disabled={disabled}
        className={cn(
          getButtonClasses({ size, wide, outline, variant, rounded, disabled, isLoading }),
          className
        )}
        {...props}
      >
        <Slottable>{children}</Slottable>
        {isLoading ? (
          <span
            className="absolute left-0 top-0 flex h-full w-full items-center justify-center bg-[inherit]"
            data-testid="loading"
          >
            <Loading
              intent={!variant || variant === "link" || variant === "ghost" ? "default" : variant}
            />
          </span>
        ) : null}
      </Comp>
    );
  }
);

Button.displayName = "Button";

export { Button };
