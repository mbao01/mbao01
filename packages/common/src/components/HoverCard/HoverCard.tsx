"use client";

import { forwardRef } from "react";
import * as HoverCardPrimitive from "@radix-ui/react-hover-card";
import { cn } from "../../utilities";
import { getHoverCardContentClasses } from "./constants";
import { HoverCardProps } from "./types";

const HoverCard = (props: HoverCardProps) => <HoverCardPrimitive.Root {...props} />;

const HoverCardContent = forwardRef<
  React.ElementRef<typeof HoverCardPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof HoverCardPrimitive.Content>
>(({ className, align = "center", sideOffset = 4, ...props }, ref) => (
  <HoverCardPrimitive.Content
    ref={ref}
    align={align}
    sideOffset={sideOffset}
    className={cn(getHoverCardContentClasses(), className)}
    {...props}
  />
));
HoverCardContent.displayName = HoverCardPrimitive.Content.displayName;

HoverCard.Content = HoverCardContent;
HoverCard.Trigger = HoverCardPrimitive.Trigger;

export { HoverCard };
