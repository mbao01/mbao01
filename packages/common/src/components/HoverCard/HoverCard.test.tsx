import { describe, expect, it } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { HoverCard } from "./HoverCard";
import { Button } from "../Button";

describe("HoverCard", () => {
  const renderHoverCard = () => {
    return render(
      <HoverCard>
        <HoverCard.Trigger asChild>
          <Button variant="link">@nextjs</Button>
        </HoverCard.Trigger>
        <HoverCard.Content>Hover card content</HoverCard.Content>
      </HoverCard>
    );
  };

  it("renders hover card trigger with hidden content", () => {
    const { asFragment } = renderHoverCard();

    expect(screen.getByRole("button", { name: "@nextjs" })).toHaveAttribute(
      "data-state",
      "closed"
    );
    expect(screen.queryByText("Hover card content")).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should show and hide hover card content", async () => {
    const user = userEvent.setup();
    renderHoverCard();
    const button = screen.getByRole("button", { name: "@nextjs" });

    expect(button).toHaveAttribute("data-state", "closed");

    await user.hover(button);

    await waitFor(() => {
      expect(screen.getByRole("button", { name: "@nextjs" })).toHaveAttribute(
        "data-state",
        "open"
      );
      expect(screen.getByText("Hover card content")).toBeInTheDocument();
    });
  });
});
