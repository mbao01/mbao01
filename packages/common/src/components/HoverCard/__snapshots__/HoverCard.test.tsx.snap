// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HoverCard > renders hover card trigger with hidden content 1`] = `
<DocumentFragment>
  <button
    class="btn disabled:text-opacity-50 btn-link min-h-fit h-10 rounded-md"
    data-state="closed"
  >
    @nextjs
  </button>
</DocumentFragment>
`;
